File: website-backend/app/routes/a/rabbit/queue_message.ts
// SPDX-License-Identifier: Apache-2.0
import express from 'express';
import { studioCors } from '@cors/studio';
import { RabbitMessage } from '@models/rabbit_message';
import { RabbitMessageStatusTypeEnum } from '@enums';
import Errors from '@root/errors';
const RequestPromise = require('request-promise');
import ConnectEnsureLogin from 'connect-ensure-login';
const Router = express.Router();

Router.options('/', studioCors);

Router.post(
  '/',
  studioCors,
  ConnectEnsureLogin.ensureLoggedIn('/user/login'),
  (req: express.Request, res: express.Response) => {
    // TODO: use RabbitMessage.sendNewMessageToQueueWithData
    let rabbitMessageId = 0;
    RequestPromise({
      method: 'POST',
      uri: 'https://x.3xr.com/messages/queue',
      form: {
        data: JSON.stringify(req.body.data),
        queue: req.body.queue,
        userId: req.user?.id,
      },
    })
      .then((xServerResult: any) => {
        rabbitMessageId = JSON.parse(xServerResult).rabbit_message_id;
        return RabbitMessage.findByPk(rabbitMessageId);
      })
      .then((rabbitMessage: any) => {
        if (rabbitMessage) {
          rabbitMessage.status_id = RabbitMessageStatusTypeEnum.SENT;
          return rabbitMessage.save();
        }
        return null;
      })
      .then((saved: any) => {
        res.json(rabbitMessageId);
      })
      .catch((err: Error) => {
        Errors.resJson(res, err, 'Error sending message');
      });
  },
);

module.exports = Router;

File: python-http/rabbit/queue_message.py
# SPDX-License-Identifier: Apache-2.0
# Queue a new rabbit message
import json
import requests
import sys

queue = sys.argv[1]
print('Sending rabbit message to queue ' + queue)
dataPayloadString = sys.argv[2]
url = 'https://www.3xr.com/a/rabbit/queue_message_internal'
data = {
  'queue': queue,
  'data': dataPayloadString,
}
print(data)

response = requests.post(url, json=data)
if (response.ok == False):
  print('Error Queuing Rabbit Message')
else:
  print('Rabbit Message (id: ' + str(response.content) + ') Sent to ' + queue)
